gson/src/test/java/com/google/gson/functional/FormattingStyleTest.java

```diff
@@ -0,0 +1,152 @@
+    String json = gson.toJson(INPUT);
+    // Make sure the default uses LF, like before.
+    assertEquals(EXPECTED_LF, json);
+  }
+
+  @Test
+  public void testNewlineCrLf() {
+    FormattingStyle style = FormattingStyle.DEFAULT.withNewline("\r\n");
+    Gson gson = new GsonBuilder().setPrettyPrinting(style).create();
+    String json = gson.toJson(INPUT);
+    assertEquals(EXPECTED_CRLF, json);
+  }
+
+  @Test
+  public void testNewlineLf() {
+    FormattingStyle style = FormattingStyle.DEFAULT.withNewline("\n");
+    Gson gson = new GsonBuilder().setPrettyPrinting(style).create();
+    String json = gson.toJson(INPUT);
+    assertEquals(EXPECTED_LF, json);
+  }
+
+  @Test
+  public void testNewlineCr() {
+    FormattingStyle style = FormattingStyle.DEFAULT.withNewline("\r");
+    Gson gson = new GsonBuilder().setPrettyPrinting(style).create();
+    String json = gson.toJson(INPUT);
+    assertEquals(EXPECTED_CR, json);
+  }
+
+  @Test
+  public void testNewlineOs() {
+    FormattingStyle style = FormattingStyle.DEFAULT.withNewline(System.lineSeparator());
+    Gson gson = new GsonBuilder().setPrettyPrinting(style).create();
+    String json = gson.toJson(INPUT);
+    assertEquals(EXPECTED_OS, json);
+  }
+
+  @Test
+  public void testVariousCombinationsToString() {
+    for (String indent : TEST_INDENTS) {
+      for (String newline : TEST_NEWLINES) {
+        FormattingStyle style = FormattingStyle.DEFAULT.withNewline(newline).withIndent(indent);
+        Gson gson = new GsonBuilder().setPrettyPrinting(style).create();
+        String json = gson.toJson(INPUT);
+        assertEquals(buildExpected(newline, indent), json);
+      }
+    }
+  }
+
+  @Test
+  public void testVariousCombinationsParse() {
+    // Mixing various indent and newline styles in the same string, to be parsed.
+    String jsonStringMix = "[\r\t'v1',\r\n        'v2'\n]";
+
+    String[] actualParsed;
+    // Test all that all combinations of newline can be parsed and generate the same INPUT.
+    for (String indent : TEST_INDENTS) {
+      for (String newline : TEST_NEWLINES) {
+        FormattingStyle style = FormattingStyle.DEFAULT.withNewline(newline).withIndent(indent);
+        Gson gson = new GsonBuilder().setPrettyPrinting(style).create();
+
+        String toParse = buildExpected(newline, indent);
+        actualParsed = gson.fromJson(toParse, INPUT.getClass());
+        assertArrayEquals(INPUT, actualParsed);
+
+        // Parse the mixed string with the gson parsers configured with various newline / indents.
+        actualParsed = gson.fromJson(jsonStringMix, INPUT.getClass());
+        assertArrayEquals(INPUT, actualParsed);
+      }
+    }
+  }
+
+  @Test
+  public void testStyleValidations() {
+    try {
+      // TBD if we want to accept \u2028 and \u2029. For now we don't.
+      FormattingStyle.DEFAULT.withNewline("\u2028");
+      fail("Gson should not accept anything but \r and \r for newline");
```

Comment:
I think these messages should say
  ...anything but \\r and \\n...
So double backslash, and the second one should be n rather than r.

Reply:
Done